### PR TITLE
Proximity features

### PR DESCRIPTION
**Pull request by catt-stefano**
Lots of stuff:

- proximity driven room info in snackbar
- snackbar for missing permissions
- geofence notification (debug mode also available)
- tracking of notifications (seen and tapped)

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -124,8 +124,12 @@
       android:name="net.squanchy.notification.NotificationsIntentService"
       android:exported="false" />
 
+    <service
+      android:name=".notification.NearNotificationIntentService"
+      android:exported="false" />
+
     <receiver
-      android:name="it.near.sdk.recipes.background.NearItBroadcastReceiver"
+      android:name=".notification.NearNotificationReceiver"
```
Comment (rock3r) [inline app/src/main/AndroidManifest.xml]: Can you undo this and use the full package name in the one you just added please?

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 2]
```diff
@@ -109,20 +126,48 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         Intent intent = getIntent();
         selectPageFrom(intent, Optional.fromNullable(savedInstanceState));
 
-        HomeComponent homeComponent = HomeInjector.obtain(this);
+        GoogleApiClient googleApiClient = new GoogleApiClient.Builder(this)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: Extract a method for the Google client init maybe?

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 3]
```diff
@@ -157,6 +202,13 @@ private void handleProximityEvent(ProximityEvent proximityEvent) {
         }
     }
 
+    private void trackProximityEnagement(Intent intent) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: This should be closer to where it's used, right after `selectPageFrom`

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 4]
```diff
@@ -313,24 +362,62 @@ public void requestSignIn() {
 
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
-        if (requestCode == REQUEST_SIGN_IN_MAY_GOD_HAVE_MERCY_OF_OUR_SOULS) {
-            startAllSubscriptions();
-        } else {
-            super.onActivityResult(requestCode, resultCode, data);
+        boolean handled = proximityPreconditions.onActivityResult(requestCode, resultCode, data);
+        if (!handled) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: Prefer positive checks:

```java
if (proximityPreconditions.onActivityResult(requestCode, resultCode, data)) {
    return;
}
```

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 5]
```diff
@@ -313,24 +362,62 @@ public void requestSignIn() {
...
     private void startAllSubscriptions() {
         subscriptions.add(
                 proximityService.observeProximityEvents()
+                        .distinctUntilChanged()
                         .observeOn(AndroidSchedulers.mainThread())
                         .subscribe(this::handleProximityEvent));
 
+        subscriptions.add(
+                proximityFeature.enabled()
+                .doOnSuccess(enabled -> {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: Seems missing a `cmd-alt-l`

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 6]
```diff
@@ -352,4 +439,57 @@ private void disposeAllSubscriptions() {
...
+        };
+    }
+
+    private void openLocationSettings() {
+        try {
+            navigator.toLocationSettingsForResult(REQUEST_SETTINGS);
+        } catch (ActivityNotFoundException e) {
+            Timber.e(e, "Unable to open location settings");
+        }
+    }
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: Almost 500 lines... we'll need to refactor this soon

File: app/src/main/java/net/squanchy/home/HomeActivity.java [hunk 7]
```diff
@@ -352,4 +439,57 @@ private void disposeAllSubscriptions() {
...
+            }
+
+            @Override
+            public void recheckAfterActivityResult() {
+                proximityPreconditions.needsActionToSatisfyPreconditions();
+            }
+        };
+    }
+
+    private void openLocationSettings() {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/home/HomeActivity.java]: Seems unused, can we remove it?

File: app/src/main/java/net/squanchy/notification/NearNotificationReceiver.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        ComponentName comp = new ComponentName(context.getPackageName(),
+                NearNotificationIntentService.class.getName());
+
+        // Start the service, keeping the device awake while it is launching.
+        startWakefulService(context, intent.setComponent(comp));
+    }
+}
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/notification/NearNotificationReceiver.java]: Missing empty line here

**Review by rock3r — CHANGES_REQUESTED** [review]
Nice one! Few small things to tweak then good to go. BTW all the `ic_notificationdroid.png`s seem unused, can we remove them?

File: app/src/main/java/net/squanchy/home/HomeActivity.java (hunk 6 above)
Comment (catt-stefano) [reply app/src/main/java/net/squanchy/home/HomeActivity.java]: I second that

**Review by rock3r — APPROVED** [review]

**Pull request merged**